### PR TITLE
feat: add discounts allocator function template (Rust, JS/TS, WASM)

### PR DESCRIPTION
**Pull request by mathiusj**
### What are we solving?

Part of https://github.com/Shopify/core-issues/issues/64459

### What was done

- added a discounts allocation function template for Rust, JS/TS and WASM to be used in new dev docs for upcoming feature

### 🎩 Instructions

- `spin up discount-functions-testing:legacy -c partners.branch=mathiusj/gh-64459-add-discounts-allocator-extension`
For each Programming Language (JS, TS, Rust, WASM)
- open discount functions testing app from that spin instance
- ` npm run shopify app generate extension -- --template discounts_allocator --name discounts-allocator-test --clone-url https://github.com/Shopify/function-examples\#mathiuss/add-discounts-allocator`
- cd into directory and run tests with either `yarn && yarn test` or `cargo build && cargo test`

### CODE REVIEW

File: discounts/javascript/discounts-allocator/default/src/run.graphql.liquid [hunk 1]
```diff
@@ -0,0 +1,7 @@
+query RunInput {
+  shop {
+    metafield(namespace: "$app:{{handle | replace: " ", "-" | downcase}}", key: "allocator-function") {
```
Comment (jharlap) [inline discounts/javascript/discounts-allocator/default/src/run.graphql.liquid:3]: for other functions we've been using the key `function-configuration`, which I think makes the intent clearer

File: discounts/rust/discounts-allocator/default/src/run.graphql.liquid [hunk 2]
```diff
@@ -0,0 +1,7 @@
+query Input {
+  shop {
+    metafield(namespace: "$app:{{handle | replace: " ", "-" | downcase}}", key: "allocator-function") {
```
Comment (jharlap) [inline discounts/rust/discounts-allocator/default/src/run.graphql.liquid:3]: same comment re: key

**Review by jharlap — CHANGES_REQUESTED** [review]
tiny change requested

File: discounts/javascript/discounts-allocator/default/src/run.graphql.liquid (hunk 1 above)
Comment (mathiusj) [reply discounts/javascript/discounts-allocator/default/src/run.graphql.liquid]: Done!

**Review by samihibrahim — COMMENTED** [review]
Why do we have different `wasm` contents compared to other discount functions? 🤔 
Other functions only have `locales` dir, `run.graphql.liquid`,  `schema.graphql` and `shopify.extension.toml.liquid`

**Review by jharlap — APPROVED** [review]

**Review by samihibrahim — APPROVED** [review]
🎩 and got all sample functions 🎉

**Pull request merged**